building-and-packaging/material/packaging_python_slides.md

```diff
@@ -0,0 +1,256 @@
+---
+type: slide
+slideOptions:
+  transition: slide
+  width: 1400
+  height: 900
+  margin: 0.1
+---
+
+<style>
+  .reveal strong {
+    font-weight: bold;
+    color: orange;
+  }
+  .reveal p {
+    text-align: left;
+  }
+  .reveal section h1 {
+    color: orange;
+  }
+  .reveal section h2 {
+    color: orange;
+  }
+</style>
+
+# How to prepare a Python code for packaging
+
+---
+
+## What is PyPI?
+
+- [PyPI](https://pypi.org/) = **The Python Package Index** is a repository of software developed in the Python community
+- PyPI itself is developed on GitHub through another software called [Warehouse](https://github.com/pypa/warehouse)
+- PyPI has a informative [public dashboard](https://p.datadoghq.com/sb/7dc8b3250-85dcf667bd?from_ts=1636972832484&to_ts=1636976432484&live=true) to show its activity
+- Who maintains PyPI and pip? A big community! For example check out [Donald Stufft](https://twitter.com/dstufft/)
+- A major advantage is the active maintenance of PyPI, which is fueled by a big community. PyPI is also on [Twitter](https://twitter.com/PyPI)
+
+---
+
+## Example of PyPI package: fenicsprecice
+
+- Having a look at [fenicsprecice](https://pypi.org/project/fenicsprecice/)
+
+---
+
+## What is pip?
+
+- pip is sometimes referred to as a *package installer* and sometimes as a *package management system*
+- For all practical purposes it is an installer for packages available at PyPI
+- pip is itself a [package](https://pypi.org/project/pip/) which is available on PyPI
+- pip is open-source and is developed on [GitHub](https://github.com/pypa/pip)
+- pip has many more things which are worth exploring! [Look here!](https://pip.pypa.io/en/latest/development/)
+
+---
+
+## Python Enhancement Proposals (PEPs)
+
+- PEP is an evolving deisgn document which provides information regarding new features of Python, new processes and new environments
+- PEPs typically involve concise technical information which also acts as standardizations
+- Packaging workflows are also standardized in through multiple PEPs. Examples are:
+    - [PEP 427](https://www.python.org/dev/peps/pep-0427/) introduces the built-package format "wheel"
+- It is always a good idea to look for **PEP** conventions before trying out processes
+
+---
+
+## Questions?
+
+---
+
+## File structure for code packaging 1/2
+
+- Most essential component of packaging is file structure. Basic file structure is as follows:
+
+```bash
+package_name/
+└── src/
+    └── solver/
+        ├── __init__.py
+        └── solver.py
+```
+
+- The file `__init__.py` is required to import the `src/` as a package. This file is mostly empty
+- `solver.py` will contain the code developed by the user
```

Comment:
I see the mistake more as having a redundant `src/` layer. A not-so-hard nomenclature is to have the source folder name the same as the main solver file name. This is what we follow in `fenicsprecice`, we have `fenicsprecice/fenicsprecice.py` there.

Reply:
But solver is here really just an example, right? That was not clear to me when reading. What would be a good generic name?

Reply:
I would have said something like `modulename` would be a generic name. At least with the structure in the slides one uses this solver package as `from solver import solver`. I guess when one uses `modulename` for the directory one would use `from modulename import solver`. In my opinion the latter way of importing the package/module is nicer since it also makes clear where the file and where the directory name play a role.

Reply:
Maybe it should rather be

```text
repository_name/
└── src/
    └── package_name/
        ├── __init__.py
        └── solver.py
```
to avoid confusstion about the package and module names? I assume a package could have several modules.

Reply:
The setuptools documentation even shows [a nice example](https://setuptools.pypa.io/en/latest/userguide/quickstart.html) that rather lookes like
```
package_name/
└── src/
    └── package_name/
        ├── __init__.py
        └── solver.py
```

The setuptools example also uses a `setup.cfg` and avoids the `setup.py`. Is the `setup.py` the old style that should be deprecated in the future? If so, it would maybe make sense to set up the installation process once with a `setup.py` and once with a `setup.cfg` in the exercise so one know both ways, but especially knows the new (=good/better) way of defining the installation procedure.

Reply:
`solver/` is just an example in this case yes. `modulename` is more generic than solver for sure.

Reply:
Isn't it bad practice that your package and module have the same name? I find `from solver import solver` ambiguous. It also confused me a bit when doing the exercise. Why not use the term `package_name` for the directory? It makes the whole slide more verbose to me. I see which part of the directory structure defines the package name. 

Using the current naming: If I just write `import solver` would it import everything inside `solver/` or does it import `solver/solver.py`? Maybe it is clear to someone who is an expert to Python, but to me it is not immediately clear.

Reply:
I am going with

```bash
package_name/
└── src/
    └── package_name/
        ├── __init__.py
        └── solver.py
```